### PR TITLE
Loging

### PR DESCRIPTION
**Pull request by edgarmagana**
login implemented and tested  using JWT, also  was tested in get users endpoint.
I tried to avoid the  IDE indentation  but not success sorry for the inconvenient

### CODE REVIEW

**Comment by reymus** [comment]
Nice job!

Here is my feedback:
- database/create_db.sql 
  - We need a separate table to store the password data.
- /controllers/login.js
  - You mixed checking for correct credentials and the token creation. A cleaner approach would be to check credentials, then create the token. Two steps in a function named `authenticate`. `createToken` is just not a good name for the authentication code (think about debugging something you didn't write.)
  - When the credentials are not valid you should throw an exception. Don't set a boolean to false.
  - 'qwerty'? Move this to `connfig.json` in a new `jwt` section.
  - Move also the expiration time to `config.json` (and set the default to 4 hours, for dev.) Set the time expressed in human readable format, `"4h"`, instead of `4 * 60 * 60`. Both mean the same thing, but `"4h"` doesn't require math ;-)
  - Use the full user object for the payload to `jwt.sign`. You are just giving it the `firt_name` which would be pretty useless. The email or user id would be good enough if the full object wasn't an option, but since it is an option, let's just serialize the full user.
- /middlewares/authenticathion.js
  - Rename the directory, from `middlewares` to `middleware` (without the 's'.) The plural of "middleware" is "middleware" (like software.)
  - Another typo! You have an extra 'h' in `authenticathion.js`. I say you just rename it to `auth.js`.
  - Yet another rename... Rename `authorization` (the function) to `authenticate`. It is a function, so it shouldn't be a noun. Additionally, authorization is a different thing we will be working on in later stages. Basically, authentication refers to the process of verifying the users are who they say they are and authorization refers to the process of verifying whether or not the users are allowed to perform certain actions (yes, I know who you are, but you are not allowed to do this.) Related: HTTP 401 vs 403.
  - If no token is found, just respond with 401.
  - If `jwt.verify` answers with an error, we need to check if the token expired and other error conditions. The invalid token error has to be the last resort. 
- /routes/users.js
  - It seems like you only added the auth middleware to the GET users endpoint. It should be added to all endpoints. I think you might just need to add it to the root express router in `routes/index.js`. Be sure to exclude the login endpoint itself, othersie, it would be impossible to login (like getting a job without experience.)
  - You need to check the parameters are provided when creating a user. You are just using `req.body.password` without verifying the parameter is set.
  - Don't use a "success" string as return value. You need to return the resource that was created. (That's one of the reasons to leave the password out of the users table.) Remember, when something goes wrong, you throw an exception. Returning a value means everything is OK, even for null values.
  - Remove console.log calls. And remember to not use them. Use proper logger calls and don't rely on console messages for debugging. Use breakpoints.
  - If user does not equal to `'success'`, the endpoint does nothing.
- /messages/errorMessages.js
  - There's a typo in `WRONG_USER_PASSWOR`. Also, change `User` for `user email` and `Password` for `password`. Move both messages to be next to each other (for maintainability purposes.)

- /util/index.js
  - Rename `encryptData` to `encrypt`
  - You should not use the sync version (hashSync.) I forgive you for now, but we will be changing this as a follow up (as an excersice in Promise/sync/await adoption, so you might escape from doing it yourself)
  - Always avoid `!=` and use `!==` instead.

**Comment by reymus** [comment]
Please also update `create_database.sql` to include 3 statements like:

```
INSERT INTO `user_passwords` VALUES(1, 'ehrgewuifhuifeo');
INSERT INTO `user_passwords` VALUES(2, 'ehrgewuifhuifeo');
INSERT INTO `user_passwords` VALUES(3, 'ehrgewuifhuifeo');
```
Where `ehrgewuifhuifeo` is the hashed value of 'admin' (it isn't, I just typed randomly) or some other simple password for our test users. Other wise, we would need to delete the users and create new ones using the POST endpoint. Additionally, a PATCH user endpoint is also needed, but we can leave that for later.

**Comment by edgarmagana** [comment]
comments attended

File: database/create_db.sql [hunk 1]
```diff
@@ -121,7 +121,14 @@ CREATE TABLE `comments` (
   CONSTRAINT `fk_comments_id_user` FOREIGN KEY (`created_by`) REFERENCES `users` (`id`) ON DELETE CASCADE
 ) ENGINE=InnoDB AUTO_INCREMENT=1 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
 
-
+CREATE TABLE `passwords` (
+  `id` int NOT NULL AUTO_INCREMENT,
```
Comment (reymus) [inline database/create_db.sql:125]: No need for an id in this table. We have a one to one relationship here, so we can have the user id as primary key.

File: server/controllers/login.js [hunk 2]
```diff
@@ -1,30 +1,62 @@
...
-    let email = userData.email;
+const authenticate = async(userData) => {
+    let user = await validateCredentials(userData);
+    let token = await createToken(user);
+    return token;
+};
+
+const validateCredentials = async(data) => {
+    let data_password = data.password;
+    let email = data.email;
```
Comment (reymus) [inline server/controllers/login.js:18]: this two variables are not needed. Avoid using variable names with "snake case." Use camel case.

File: server/controllers/login.js [hunk 3]
```diff
@@ -1,30 +1,62 @@
...
-        let match = await bcrypt.compare(password, user.password);
+        let password = await getPassword(user.id);
+        let match = await bcrypt.compare(data_password, password);
 
-        if (!match) {
-            return {
-                token: false,
-                message: messages.WRONG_USER_PASSWOR,
-            };
+        if (match === false) {
```
Comment (reymus) [inline server/controllers/login.js:25]: Use `!match`

File: server/controllers/login.js [hunk 4]
```diff
@@ -1,30 +1,62 @@
...
+    let token = await createToken(user);
+    return token;
+};
+
+const validateCredentials = async(data) => {
+    let data_password = data.password;
+    let email = data.email;
 
     try {
         let user = await usersController.getUserByEmail(email);
```
Comment (reymus) [inline server/controllers/login.js:21]: You should fetch the user before entering this function. `validateCredentials` should receive the user and the data to validate against (the userData you are using.) The function has to return `true` or `false` NOT a user object. Look at it from the API user perspective... Why would validateCredentials return a user?

In the end, the `authenticate` function should look like this:

```
const authenticate = async(userData) => {
    let user = await usersController.getUserByEmail(userData.email);
    let valid = await validateCredentials(user, userData.password);
    if (valid) {
        let token = await createToken(user);
        return token;
    } else {
        let message = messages.WRONG_USER_PASSWOR;
        throw new Error(message);
    }
};
```

File: server/controllers/login.js [hunk 5]
```diff
@@ -1,30 +1,62 @@
...
+    try {
+        let token = jwt.sign({ user }, config.jwt.secret, { expiresIn: config.jwt.expire });
+        return token;
+    } catch (error) {
+        let message = messages.TOKEN_UNAVAILABLE
+        throw new Error(message);
+    }
+};
+
+const getPassword = async(id) => {
```
Comment (reymus) [inline server/controllers/login.js:48]: Rename `id` to `userId`

File: database/create_db.sql [hunk 6]
```diff
@@ -121,7 +121,14 @@ CREATE TABLE `comments` (
...
+  `id_user` int NOT NULL,
```
Comment (reymus) [inline database/create_db.sql:126]: `id_user` should be either `user` or `user_id`

File: server/controllers/login.js [hunk 7]
```diff
@@ -1,30 +1,62 @@
...
+    try {
+        let queryModel = new QueryModel(Passwords);
+        let query = queryModel.select("password").where('id_user').equals(id).build();
+        let result = await db.query(query);
+        return result[0].password;
```
Comment (reymus) [inline server/controllers/login.js:53]: Verify the result has at least `length === 1`

File: server/controllers/users.js [hunk 8]
```diff
@@ -44,6 +56,23 @@ const createUser = async(user) => {
     }
 }
 
+const insertPassword = async(val) => {
```
Comment (reymus) [inline server/controllers/users.js:59]: Rename this to `createPassword`

Comment (reymus) [reply server/controllers/users.js:59]: This function has to receive a user or user id and a password. Receiving an array is a bad design.

File: server/controllers/users.js [hunk 9]
```diff
@@ -20,19 +20,31 @@ const getUserByEmail = async(email) => {
 const createUser = async(user) => {
     let model = Users;
     let values = [];
+    let passwordValues = [];
     let fields = Object.keys(model.fields);
+
```
Comment (reymus) [inline server/controllers/users.js:25]: Extract everything related to passwords from this function. Make a new function `createUserWithPassword` and make it invoke `createUser` and `createPassword`

File: server/middleware/auth.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+            if (err) {
+                res.status('401').json({
+                    message: messages.INVALID_TOKEN
+                });
+            }
+            req.user = decoded.user;
+            next();
+        });
+    }
+    next();
```
Comment (reymus) [inline server/middleware/auth.js:21]: This needs an `else` or the `next()` above needs a `return`

File: server/middleware/auth.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    if (!config.validation.off) {
+
+        let token = req.get('Authorization');
+        jwt.verify(token, config.jwt.secret, (err, decoded) => {
+            if (err) {
+                res.status('401').json({
+                    message: messages.INVALID_TOKEN
+                });
+            }
+            req.user = decoded.user;
```
Comment (reymus) [inline server/middleware/auth.js:17]: This has to be in an `else`

File: server/routes/users.js [hunk 12]
```diff
@@ -14,12 +14,12 @@ app.get('/', authorization, async(req, res) => {
     }
 });
 
-app.post('/', async(req, res) => {
+app.post('/', authenticate, async(req, res) => {
 
     try {
-        let encrypted = encryptData(req.body.password);
+        let encrypted = encrypt(req.body.password);
         //validate email is missing
```
Comment (reymus) [inline server/routes/users.js:21]: What is this comment doing here?

File: server/controllers/login.js (hunk 3 above)
Comment (edgarmagana) [reply server/controllers/login.js:25]: for some reason that not works yesterday (I don't know why) but let me try again

File: server/controllers/users.js (hunk 8 above)
Comment (edgarmagana) [reply server/controllers/users.js:59]: done

File: database/create_db.sql (hunk 1 above)
Comment (edgarmagana) [reply database/create_db.sql:125]: done

File: database/create_db.sql (hunk 6 above)
Comment (edgarmagana) [reply database/create_db.sql:126]: done

File: server/controllers/login.js (hunk 5 above)
Comment (edgarmagana) [reply server/controllers/login.js:48]: done

File: server/controllers/login.js (hunk 7 above)
Comment (edgarmagana) [reply server/controllers/login.js:53]: done

File: server/controllers/login.js (hunk 2 above)
Comment (edgarmagana) [reply server/controllers/login.js:18]: done

File: server/controllers/login.js (hunk 3 above)
Comment (edgarmagana) [reply server/controllers/login.js:25]: done

File: server/middleware/auth.js (hunk 11 above)
Comment (edgarmagana) [reply server/middleware/auth.js:17]: done

File: server/middleware/auth.js (hunk 10 above)
Comment (edgarmagana) [reply server/middleware/auth.js:21]: done

File: server/controllers/login.js (hunk 4 above)
Comment (edgarmagana) [reply server/controllers/login.js:21]: done

File: server/routes/users.js (hunk 12 above)
Comment (edgarmagana) [reply server/routes/users.js:21]: sorry  it was validate email is needed

File: server/controllers/users.js (hunk 9 above)
Comment (edgarmagana) [reply server/controllers/users.js:25]: not sure about that I think the correct way is use a transaction so,  probably is better to do this as improvement( paco's work ;) )

Comment (edgarmagana) [reply server/controllers/users.js:20]: done and  pushed

File: database/create_db.sql [hunk 13]
```diff
@@ -17,6 +17,7 @@ CREATE TABLE `users` (
   `first_name` varchar(45) NOT NULL,
   `last_name` varchar(45) NOT NULL,
   `email` varchar(45) NOT NULL,
+  `password` varchar(255) NOT NULL,
```
[the remark this answers is not in the document]
Comment (edgarmagana) [reply database/create_db.sql]: done

File: server/messages/errorMessages.js [hunk 14]
```diff
@@ -25,5 +29,9 @@ const errors = (code) => {
 }
 module.exports = {
     NOT_FOUND,
+    PASSWORD_REQUIRED,
+    WRONG_USER_PASSWOR,
```
[the remark this answers is not in the document]
Comment (edgarmagana) [reply server/messages/errorMessages.js:33]: done

File: server/middleware/auth.js [hunk 15]
```diff
@@ -0,0 +1,29 @@
+const jwt = require('jsonwebtoken');
+const messages = require('../messages/errorMessages');
+const config = require('./../config')
+
+
+let authenticate = (req, res, next) => {
+    // delte if for validate the token  this is  to prevent Paco have errors to consume the endpoints
```
[the remark this answers is not in the document]
Comment (edgarmagana) [reply server/middleware/auth.js:7]: done

File: server/controllers/login.js [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+};
+
+const getPassword = async(userId) => {
+    try {
+        let queryModel = new QueryModel(Passwords);
+        let query = queryModel.select("password").where('user_id').equals(userId).build();
+        let result = await db.query(query);
+        if (result.length === 1) {
+            return result[0].password;
+        }
```
[the remark this answers is not in the document]
Comment (edgarmagana) [reply server/controllers/login.js:53]: done

File: server/controllers/users.js [hunk 17]
```diff
@@ -17,19 +17,30 @@ const getUserByEmail = async(email) => {
...
         password = "'" + user.password + "'";
-    } else {
-        let message = messages.PASSWORD_REQUIRED;
+        let userResult = await createUser(user);
+        userId = userResult.insertId;
+        let passwordResult = await createPassword(password, userId);
+        if (passwordResult.affectedRows === 1) {
+            //return user created is needed
+            return 'success';
+        }
```
[the remark this answers is not in the document]
Comment (edgarmagana) [reply server/controllers/users.js:31]: done

File: server/controllers/users.js [hunk 18]
```diff
@@ -38,22 +49,19 @@ const createUser = async(user) => {
...
-        passordResult = await createPassword(password, userId);
     } catch (e) {
         let message = messages.errors(e.errno);
         throw new Error(message);
     }
 
     if (result.insertId != null) {
         //return the object created
-        return 'success';
+        return result;
```
[the remark this answers is not in the document]
Comment (edgarmagana) [reply server/controllers/users.js:64]: done

File: server/controllers/users.js [hunk 19]
```diff
@@ -38,22 +49,19 @@ const createUser = async(user) => {
...
     }
```
[the remark this answers is not in the document]
Comment (edgarmagana) [reply server/controllers/users.js:68]: done

File: server/routes/users.js [hunk 20]
```diff
@@ -27,7 +27,7 @@ app.post('/', authenticate, async(req, res) => {
             return;
         }
         req.body.password = encrypted;
-        let user = await controller.createUser(req.body);
+        let user = await controller.createUserWithPassword(req.body);
         if (user === 'success') {
             res.status(201).send(user);
         }
```
[the remark this answers is not in the document]
Comment (edgarmagana) [reply server/routes/users.js:33]: done

File: server/routes/users.js [hunk 21]
```diff
@@ -1,14 +1,47 @@
...
+            });
+            return;
+        }
+        req.body.password = encrypted;
+        let user = await controller.createUserWithPassword(req.body);
+        if (user === 'success') {
+            res.status(201).send(user);
+        } else {
+            res.status('500').json({
+                message: err.message,
```
Comment (reymus) [inline server/routes/users.js:35]: err is not defined.

File: server/controllers/login.js [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+        let query = queryModel.select("password").where('user_id').equals(userId).build();
+        let result = await db.query(query);
+        if (result.length === 1) {
+            return result[0].password;
+        } else {
+            let message = messages.UNABLE_TO_OBTAIN;
+            throw new Error(message);
+        }
+    } catch (error) {
+        let message = messages.errors(e.errno);
```
Comment (reymus) [inline server/controllers/login.js:58]: e is not defined
